File: python/2015/08/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
sys.path.append("../")
import util as u
import re

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def analyze(a):
    l = len(a)
    a = a.replace("\\\\","*")
    a = a.replace("\\\"","+")
    a = re.sub("\\\\x[0-9a-f][0-9a-f]","-", a)
    return (l,len(a)-2)

def encode(a):
    l = len(a)
    a = a.replace("\\", "\\\\")
    a = a.replace("\"", "\\\"")
    return (len(a)+2, l)

def read_input():
    return u.read_lines(file)

def part1():
    A = read_input()
    l = [analyze(a) for a in A]
    print("Part 1:",sum([i[0] for i in l]) - sum([i[1] for i in l]))

def part2():
    A = read_input()
    l = [encode(a) for a in A]
    print("Part 2:",sum([i[0] for i in l]) - sum([i[1] for i in l]))

part1()
part2()

File: c/2015/08/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif

int16_t unescape(char* str) {
    int16_t index = 0;
    for (int16_t i = 1; i < strlen(str)-1; i++) {
        if (str[i] != '\\') {
            index++;
        } else {
            if (str[i+1] == '\\') {
                index++;
                i++;
            } else if (str[i+1] == '"') {
                index++;
                i++;
            } else {
                index++;
                i+=3;
            }
        }
    }
    return index;
}

int16_t escape(char* str) {
    int16_t index = 2;
    for (int16_t i = 0; i < strlen(str); i++) {
        if (str[i] == '"' || str[i] == '\\') index++;
        index++;
    }
    return index;
}

void part1() {
    ARR_STRING* in = readInput(FILE_NAME);
    int64_t k = 0;
    for (int16_t i = 0; i < in->length; i++) {
        k += strlen(in->strings[i]);
        k -= unescape(in->strings[i]);
    }
    freeArrString(in);
    printf("Part 1: %ld\n", k);
}

void part2() {
    ARR_STRING* in = readInput(FILE_NAME);
    int64_t k = 0;
    for (int16_t i = 0; i < in->length; i++) {
        k -= strlen(in->strings[i]);
        k += escape(in->strings[i]);
    }
    printf("Part 2: %ld\n", k);
    freeArrString(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}
